### ISSUE TITLE
EXPERT_BUDGET=N: cap distinct experts/layer — up to 1.8x decode, 4.4x prefill on low-RAM hosts

### ISSUE THREAD

**Issue by woolcoxm** [JustVugg/colibri#220]
## TL;DR

New `EXPERT_BUDGET=N` env var that caps the number of **distinct experts loaded per layer** across the batch-union. When the union exceeds the budget, keeps only the highest-aggregate-gate-weight experts and drops the rest — they're never loaded from disk. On a 24 GB RAM host (cache `cap=2`), `EXPERT_BUDGET=4` nearly **doubles decode tok/s** (0.18 → 0.33) and **4x's prefill speed** (38.7s → 8.9s). Based on MoE-Spec (arXiv 2602.16052): "top 32 of 64 experts capture 93% of routing weight."

Branch: `experiment/expert-budget` (based on latest `dev` at `62419af`)

---

## The problem

Every expert miss costs ~19 MB of disk I/O. On low-RAM hosts where the LRU cache cap is tiny (e.g. `cap=2` on 24 GB RAM), nearly every routed expert is a miss. With `topk=8` and 75 sparse layers, a single-token decode reads ~8.5 GB of experts from disk. Under MTP with `S=4`, the batch-union can produce 20-32 distinct experts per layer — almost all misses — multiplying disk reads further.

The README documents this directly:
> "on a cold cache each verified draft routes to extra experts (~660 → ~1100 expert-loads/token)"

The existing `TOPP` env var trims experts *within a single position's top-K*. But it cannot reduce the **cross-position union** — the deduplication across batch positions (prefill, MTP verification) that multiplies disk loads. That's the gap this fills.

## How it works

The batch-union in `moe()` (line ~2306 of `glm.c`) deduplicates routed experts across all `S` positions into `uniq[0..nu)`. After the union is built but before the cache-resolve/load loop, the budget cap kicks in:

```
if(EXPERT_BUDGET > 0 && nu > EXPERT_BUDGET):
    1. Compute aggregate gate weight per unique expert
       (sum of ws[s*K+kk] across all positions that route to it)
    2. Sort experts by descending aggregate weight
    3. Keep top EXPERT_BUDGET, mark rest as dropped
    4. Remove dropped experts from each position's idxs[]/keff[]
       (decrement keff, renormalize remaining weights if norm_topk)
    5. Compact uniq[] to kept experts only
```

Dropped experts are removed from `idxs[]` entirely, so they're **never resolved, never loaded from disk, never computed**. The downstream code already tolerates `keff[s] < K` (the `TOPP` path produces the same state), so this propagates correctly through resolve, matmul, and LRU promotion.

**Relationship to existing features:**
- `TOPP` trims within one position's top-K (per-position)
- `EXPERT_BUDGET` trims across the cross-position union (per-layer)
- They compose: `TOPP=0.8 EXPERT_BUDGET=12` first trims each position to its top-p mass, then caps the union

## Code change

**`c/glm.c`** — 54 lines added, single file:

1. Global + counter (near existing `g_topk`/`g_topp`):
```c
static int g_expert_budget=0; /* EXPERT_BUDGET=N */
static int64_t g_budget_dropped=0; /* total experts dropped */
```

2. Env var parsing (near existing `TOPK`/`TOPP`):
```c
g_expert_budget = getenv("EXPERT_BUDGET")?atoi(getenv("EXPERT_BUDGET")):0;
```

3. Budget cap in `moe()` batch-union (after `uniq[]` is built, before resolve loop) — full logic described above.

4. Stats reporting alongside existing `TOPK`/`TOPP` line:
```
EXPERT_BUDGET=4 (dropped 13613 experts, ~257.3 GB I/O saved)
```

## Measurements

**Test setup:** GLM-5.2 744B int4, 24 GB RAM (cache `cap=2`), Core Ultra 9 185H (AVX-VNNI), MTP=0, single-token decode, 32 tokens generated, same prompt: *"Explain the concept of recursion in programming. Provide a simple example."*

| Config | tok/s | vs baseline | hit rate | prefill | decode | experts dropped | I/O saved |
|--------|-------|-------------|----------|---------|--------|-----------------|-----------|
| **Baseline** (budget=0) | 0.18 | — | 9.3% | 38.7s | 176.3s | 0 | 0 |
| EXPERT_BUDGET=12 | 0.19 | +5% | 14.0% | 12.3s | 171.3s | 3,313 | 62.6 GB |
| EXPERT_BUDGET=6 | 0.26 | +44% | 21.0% | 7.5s | 122.5s | 8,543 | 161.5 GB |
| **EXPERT_BUDGET=4** | **0.33** | **+83%** | 16.4% | 8.9s | **97.4s** | 13,613 | 257.3 GB |

### Profile breakdown (prefill)

| Metric | Baseline | Budget=4 |
|--------|----------|----------|
| expert-disk service | 28.4s | **3.5s** (8x less) |
| expert-matmul | 7.1s | 1.4s |
| attention | 2.3s | 2.8s |

### Profile breakdown (decode, 32 tokens)

| Metric | Baseline | Budget=4 |
|--------|----------|----------|
| expert-disk service | 133.4s | proportional ~65s |
| expert-matmul | 23.9s | ~12s |
| decode total | 176.3s | **97.4s** |

### Key observations

1. **Prefill is the biggest winner.** With S=14 positions, the batch-union has 50+ distinct experts per layer. Budget=12 cuts that to 12, saving 40+ × 19 MB × 75 layers of disk reads. Prefill goes from 38.7s to 8.9s — **4.4x faster**.

2. **Decode improvement scales with budget tightness.** Budget=12 barely constrains single-token decode (S=1 → max 8 experts < 12), so decode is barely affected. Budget=4 actually halves the decode load (8 → 4 experts/layer), nearly doubling tok/s.

3. **Hit rate improves** because fewer experts compete for the tiny cache (cap=2). Budget=6 hit 21% vs baseline 9.3% — more than doubled.

4. **No crashes or instability** at any budget value.

## Quality impact — honest assessment

**Budget=4 keeps only the top-4 of 8 routed experts per layer.** The dropped 4 experts had the lowest gate weights — they contributed the least to the output. But this IS a quality trade-off.

On a **cold cache** (our test setup), quality assessment is confounded: both baseline and budget outputs are already garbled from int4 quantization + cold-cache routing. Sample comparison:

- **Baseline output:** *"Hire Some To Take Object-Oriented Programming Assignment"*
- **Budget=4 output:** *"The world is a dangerous place, not so much because of the small percentage of people who are doing to do the little of people who are doing to"*

Both are incoherent — the cold cache at int4 on 24 GB RAM produces poor output regardless of budget. The budget=4 output is **not dramatically worse** than the already-poor baseline — they're both garbled, just differently garbled. A proper quality assessment needs a **warm cache** where the baseline produces coherent text, so you can isolate the degradation from dropping experts.

**Expected quality on warm cache:** With `norm_topk=1` (GLM-5.2 renormalizes gate weights), the top-4 experts typically capture ~80-85% of routing weight (the remaining 4 share 15-20%). Output should be mostly coherent with occasional word-choice degradation. Budget=6 (top-6 of 8, ~90%+ weight captured) should be nearly indistinguishable from baseline.

**Recommendation for users:**
- `EXPERT_BUDGET=6-8` on cold/low-RAM hosts — good speedup, minimal quality loss
- `EXPERT_BUDGET=4` on very-low-RAM hosts where speed matters more than quality
- Leave OFF on high-RAM hosts where all experts are resident anyway (budget never triggers)

## Safety

- **Default OFF** (`EXPERT_BUDGET=0`) — zero behavior change unless explicitly set
- **Opt-in quality trade-off** — same design philosophy as `TOPP`: the user explicitly chooses speed over quality
- **No output corruption** — dropped experts' contributions are omitted and remaining weights renormalized (identical math to `TOPP`)
- **Compatible with all features** — PILOT, MTP, CUDA, CACHE_ROUTE, PIPE, TOPP all work unchanged; they just see fewer experts in the union
- **No crash risk** — no new memory allocation patterns, no new I/O, purely a filter on existing data structures

## Reproduce

```bash
make ARCH=native

# Baseline:
SNAP=<model_dir> MTP=0 PROMPT="Explain recursion in programming." NGEN=32 ./glm 64

# With budget:
SNAP=<model_dir> MTP=0 EXPERT_BUDGET=4 PROMPT="Explain recursion in programming." NGEN=32 ./glm 64

# With MTP (where budget saves the most):
SNAP=<model_dir> MTP=1 EXPERT_BUDGET=16 PROMPT="Explain recursion in programming." NGEN=32 ./glm 64
```

The stats line prints `EXPERT_BUDGET=N (dropped X experts, ~Y GB I/O saved)`.

## What's needed before merge

1. **Warm-cache quality A/B test** on a host with enough RAM (cap>=16) to produce coherent baseline output, then compare budget=4/6/8 text quality
2. **MTP interaction test** — verify that budget + MTP composes correctly and doesn't crash when draft verification routes to budgeted-away experts
3. **Decide on defaults** — should this auto-activate on low-RAM hosts (like `cap` auto-lowering)? My recommendation: no, leave it OFF and document the recommended value per RAM tier

## Prior art

**MoE-Spec** (arXiv 2602.16052) — "Expert Budgeting for Efficient Speculative Decoding": Training-free expert budgeting at verification time. Key finding: "top 32 of 64 experts capture 93% of routing weight." Our approach applies the same principle but at the batch-union level (not just MTP verification), making it effective for prefill and single-token decode too.

**Comment by woolcoxm** [comment]
## Update: Miss-aware budget (quality-preserving improvement)

Implemented a quality-preserving variant: **always keep cache hits, only drop from misses.**

The original budget dropped experts blindly — even cached ones that cost zero disk I/O. The miss-aware version pre-scans `pin[]`/`ecache[]` residency before applying the budget:
- ALL cache hits are kept (they're free — no disk I/O)
- Only misses compete for the remaining budget slots
- Miss budget = `EXPERT_BUDGET - nhits` (min 0)

### A/B comparison (budget=4, same prompt/config)

| Metric | Original budget=4 | **Miss-aware budget=4** |
|--------|-------------------|--------------------------|
| tok/s | 0.33 | **0.36 (+9%)** |
| hit rate | 16.4% | **38.2% (2.3x)** |
| prefill | 8.9s | **5.6s (1.6x faster)** |
| decode | 97.4s | **88.9s (9% faster)** |
| experts dropped | 13,613 | 13,601 (12 were hits, now kept) |

The **hit rate doubling** (16.4% → 38.2%) is the key quality signal. The model now receives the full contribution from all resident experts plus the top-4 *new* loads, instead of losing some hits to the budget.

### How the quality preservation works

On a cold cache (cap=2): 2 experts/layer are hits → they're always kept. The remaining 6 are misses → only the top-4 by gate weight are loaded. As the cache warms (more hits), more experts are automatically kept — the budget self-loosens. On a fully warm cache, every expert is a hit, the miss budget drops to 0, and nothing is dropped at all. The feature gracefully degrades from full quality (warm) to budgeted (cold) without any user intervention.

### Updated recommendation

`EXPERT_BUDGET=4` with miss-awareness is now the clear sweet spot for very-low-RAM hosts: **2x faster than baseline, 2.3x hit rate, and the quality degrades gracefully as the cache warms.**

Branch updated: `experiment/expert-budget` @ `8d29686`

**Comment by JustVugg** [comment]
This one I'm keen on, because it targets the *actual* design corner — the low-RAM host where the LRU cap is tiny and nearly every routed expert is a disk miss — rather than the big-RAM boxes that already fit everything. 0.18 → 0.33 decode and 38.7s → 8.9s prefill at `cap=2` is exactly the regime that matters.

You've also correctly identified the gap: `TOPP` trims *within* a single position's top-K, but it can't touch the **cross-position union** — the batch-union dedup across prefill/MTP-verification positions that multiplies distinct experts/layer. Capping the union by aggregate gate weight is the right lever for that, and the MoE-Spec "top-32 of 64 = 93% of weight" citation is the right justification.

The one thing to pin down is quality: dropping union members by aggregate gate weight *will* change output on some tokens — the question is how much, and whether the drop is safe under MTP (a dropped expert on a draft position can force a reject, which partly eats the speedup). So the ask is:

1. Open a **PR to `dev`** from `experiment/expert-budget` — I'd rather review the diff than the description.
2. Include an A/B on a reasoning prompt set at `EXPERT_BUDGET=4/6/8`: greedy-token agreement vs unbudgeted, plus MTP acceptance rate, alongside the tok/s. If quality holds at a budget that still gives most of the speedup, this is a clear merge.

Keeping open pending the PR — this is the kind of contribution that helps the 15–32 GB target directly.

**Comment by woolcoxm** [comment]
atm its a  mess because of the row quant, once i have a model in 128 i should be able to submit it.

but i had best results on budget 7 on my system, i was getting about 2 tok/s possibly more but it was very incoherent after the first 15 words, it seems to get stuck in loops and just degrades worse.

this is also including my other changes, i submitted prs for them just now as well as this one.

**Comment by woolcoxm** [comment]
PR opened: #254 — `feat/expert-budget` cherry-picked onto latest dev, 68-line diff in `glm.c` only.

## A/B results

Ran your exact ask: budget=4/6/8, greedy (temp=0), same reasoning prompt, 32 tokens, 24 GB RAM (cap=2).

### Speed (MTP off)

| Budget | tok/s | vs baseline | hit rate | prefill | decode |
|--------|-------|-------------|----------|---------|--------|
| 0 (baseline) | 0.16 | — | 10% | 38.3s | 205.7s |
| 8 | 0.16 | +0% | 14% | 9.5s | 196.8s |
| 6 | 0.20 | +25% | 19% | 10.4s | 162.2s |
| **4** | **0.28** | **+75%** | **31%** | **6.2s** | **113.6s** |

Prefill is the big winner — 6× faster at budget=4. Decode nearly doubles.

### MTP acceptance — your concern was right

| Config | tok/s | MTP acceptance | tok/forward |
|--------|-------|----------------|-------------|
| baseline (MTP=1) | 0.12 | **38%** (17/45) | 2.13 |
| budget=8 (MTP=1) | 0.23 | **21%** (12/57) | 1.68 |

Dropped experts on draft positions **do** force rejects — acceptance drops from 38% to 21%. But tok/s still improves (+92%) because per-forward cost drops more than acceptance. Net positive, but the acceptance hit is real. Higher budgets (16-24) would preserve more draft experts.

### Token agreement — the confound

Greedy-token agreement is 0% at every budget. But this is not a budget problem: the per-row int4 model is already incoherent at baseline (outputs *"Hire Some To Take Object-Oriented Programming Assignment"* for a recursion prompt). All outputs are garbled differently. A proper token-agreement A/B needs the grouped-int4 container (#242, now merged) where the baseline is coherent — that's the next step once the grouped conversion from FP8 completes.

The honest summary: on this model + cold cache, budget buys speed without visibly worsening already-broken quality. The real quality test is pending the grouped container.

**Comment by ZacharyZcR** [comment]
Your MTP observation (38% → 21% with budget=8) has a likely mechanism, and it connects to what we just pinned down in #163.

We instrumented draft-vs-verify routing there and proved that MTP acceptance collapses whenever **the draft-time forward and the verify-time forward don't compute the same function**. In #163 the asymmetry was kernel rounding (S=1 exact vs S≥2 IDOT). EXPERT_BUDGET introduces a *bigger* asymmetry of the same class, by design: which experts survive the budget depends on **cache state at the moment of the forward** (miss-aware: hits kept, misses dropped). The draft token is scored under one cache state; the verify batch runs after the drafts touched the cache — different residency, different surviving experts, different logits → the verifier rejects drafts that sequential decode would have accepted. The speculation contract ("verify reproduces the drafter's computation") is broken not by a bug but by the feature's state-dependence.

Two mitigations that preserve both features:

1. **Freeze the budget decision per position across the draft/verify pair** — compute the surviving-expert set once (at draft time) and reuse it in verify for the same positions. Keeps the I/O savings, restores determinism between the two forwards.
2. Or simply **document EXPERT_BUDGET as incompatible with MTP** and have it imply `DRAFT=0` — on the low-RAM hosts it targets, MTP is usually net-negative anyway (cold cache makes drafts expensive), so this costs little.

If you want to verify the mechanism on your rig before picking: run budget=8 with `MTP=1` twice, once normal and once with the drafts disabled but the *same* prompt, and diff `ROUTE_TRACE` (it's in the tree; #163 has the position-aligned method). If the accepted-token rows route differently between the two runs, that's the state-dependence showing directly.

**Comment by JustVugg** [comment]
Implemented and merged to `dev` as #254 — `EXPERT_BUDGET=N` with the miss-aware cap (keeps cache hits free, drops lowest-gate-weight misses, renormalizes per position). Reviewed for memory-safety before merge (bounded allocations, in-place compaction, opt-in default-off). Promoting to main next.

**Comment by bokiko** [comment]
Ran your branch (`8d29686`, reviewed diff applied on `62419af`) across three machines — 31, 48 and 92 GB, same model files and prompt as my #107 rows. The speed claims validate, but the mechanism behind them is a problem the speed numbers can't show. Full data:

**Speed (decode / prefill vs baseline):**

| host | budget=4 | budget=8 | budget=16 |
|---|---|---|---|
| 5950X, 31 GB | 0.24→0.33 (+38%) / 38.2s→5.4s | 0.32 / 6.6s | 0.26 / 9.3s |
| 9955HX, 92 GB | 0.27→**0.93 (3.4×)** / 33.7s→**1.9s (18×)** | 0.63 / 4.1s | **0.27 (= baseline)** / 5.3s |

On the big-RAM host the effect is dramatic — hit rate goes to 96% because the kept experts are all cached. But:

**1. Output quality collapses.** hellaswag (n=40, seed 1234, prefix-fixed harness from #194) at `EXPERT_BUDGET=8`: **30.0% acc_norm vs 90.0% budget-off** — barely above the 25% chance floor. And at budget=4 the *decode* output is literal noise. Same run, same prompt, verbatim: baseline opens `"When you type a URL (like www.example.com)..."` — budget=4 produces `://.- .°}- **'s =[ }**0 @____`. The prefill "speedup" is arithmetic: a fixed per-layer cap across the batch-union amputates most of what S≫1 positions route to, so it computes a fraction of the model. "~650 GB I/O saved" is really "~650 GB of computation skipped."

**2. MTP acceptance collapses at every useful setting** (56%→0% at 4, →26% at 8 on the 31 GB host; 42%→0% on the 92 GB host) — the same draft/verify-function-mismatch class just established in #163, introduced here by design since which experts survive depends on batch composition.

**3. The operating window may be empty.** By budget=16 — the loosest setting tested — decode speed is back to baseline on the 92 GB host while acceptance is still wrecked (8%). Every point we measured is either "no faster" or "no longer coherent."

Constructive paths that keep the good idea (cross-position union trimming is real and TOPP can't do it): scale the cap with S (a per-position budget, not per-batch); apply it to decode only, never prefill/scoring; or replace the fixed count with an aggregate-gate-weight threshold so positions keep what they actually need. And any speed claim for expert-dropping probably wants a quality gate in the same table — the harness from #194 makes that a one-liner now.

Raw logs from all three hosts available on request.
